src/Context.hs

```diff
@@ -121,70 +181,110 @@ replaceHistory' = flip replaceHistory
 -- In most cases the qualified path will have been qualified under the same
 -- context, but this constraint is *not* enforced by the definition of this
 -- function.
-insertInGlobalEnv :: Context -> QualifiedPath -> Binder -> Context
+insertInGlobalEnv :: Context -> QualifiedPath -> Binder -> Either ContextError Context
 insertInGlobalEnv ctx qpath binder =
-  let globalEnv = contextGlobalEnv ctx
-   in ctx {contextGlobalEnv = envInsertAt globalEnv (unqualify qpath) binder}
+  replaceLeft
+    (FailedToInsertInGlobalEnv (unqualify qpath) binder)
+    ( E.insert (contextGlobalEnv ctx) (unqualify qpath) binder
+        >>= \e -> pure $! (ctx {contextGlobalEnv = e})
+    )
 
 -- | Adds a binder to a context's type environment at a qualified path.
 --
 -- In most cases the qualified path will have been qualified under the same
 -- context, but this constraint is *not* enforced by the definition of this
 -- function.
-insertInTypeEnv :: Context -> QualifiedPath -> Binder -> Context
-insertInTypeEnv ctx qpath binder =
-  let typeEnv = getTypeEnv (contextTypeEnv ctx)
-   in ctx {contextTypeEnv = TypeEnv (envInsertAt typeEnv (unqualify qpath) binder)}
+insertTypeBinder :: Context -> QualifiedPath -> Binder -> Either ContextError Context
+insertTypeBinder ctx qpath binder =
+  let (SymPath path name) = unqualify qpath
+   in first
+        (\_ -> trace (show path) (FailedToInsertInTypeEnv (unqualify qpath) binder))
+        ( case path of
+            [] ->
+              (E.insert (contextTypeEnv ctx) (SymPath [] name) binder)
+                >>= pure . (replaceTypeEnv ctx)
+            -- TODO: We need to 'view' the global environment as a type
+            -- environment here to ensure types are added to a module's type
+            -- environment and not its value environment (the modality is
+            -- correct)
+            -- Find a more elegant API here.
+            _ ->
+              (E.insert (TypeEnv (contextGlobalEnv ctx)) (SymPath path name) binder)
+                >>= pure . (replaceGlobalEnv ctx) . getTypeEnv
+        )
+
+replaceTypeBinder :: Context -> QualifiedPath -> Binder -> Either ContextError Context
+replaceTypeBinder ctx qpath binder =
+  let (SymPath path name) = unqualify qpath
+   in first
+        (\_ -> trace (show path) (FailedToInsertInTypeEnv (unqualify qpath) binder))
+        ( (E.replaceInPlace (contextTypeEnv ctx) name binder)
```

Comment:
Stylistically, I think this would look better if it's one line and drops a bunch of parenthesis (it must surely be possible, due to the operators being used?)